modules/nextflow/src/main/groovy/nextflow/trace/TaskCleanupObserver.groovy

```diff
@@ -0,0 +1,452 @@
+                    deleteTask(task)
+                else if( strategy >= CleanupStrategy.AGGRESSIVE && canDeleteFile(source) )
+                    deleteFile(source)
+            }
+            else {
+                log.trace "File ${source.toUriString()} was published before task was marked as completed"
+
+                // save file to be processed when task completes
+                publishedOutputs << source
+            }
+        }
+    }
+
+    /**
+     * When a process is closed (all tasks of the process have been created),
+     * mark the process as closed and scan tasks for cleanup.
+     *
+     * @param process
+     */
+    @Override
+    void onProcessClose(TaskProcessor process) {
+        sync.withLock {
+            processes[process.name].closed = true
+            cleanup0()
+        }
+    }
+
+    /**
+     * When the workflow completes, delete all task directories (only
+     * when using the 'lazy' strategy).
+     */
+    @Override
+    void onFlowComplete() {
+        if( strategy == CleanupStrategy.LAZY && session.isSuccess() ) {
+            log.info 'Deleting task directories (this might take a moment)...'
+
+            for( TaskRun task : tasks.keySet() )
+                deleteTask(task)
+        }
+
+        threadPool.shutdown()
+    }
+
+    /**
+     * Delete any task directories and output files that can be deleted.
+     */
+    private void cleanup0() {
+        if( strategy < CleanupStrategy.EAGER )
+            return
+
+        for( TaskRun task : tasks.keySet() )
+            if( canDeleteTask(task) )
+                deleteTask(task)
+
+        if( strategy < CleanupStrategy.AGGRESSIVE )
+            return
+
+        for( Path path : paths.keySet() )
+            if( canDeleteFile(path) )
+                deleteFile(path)
+    }
+
+    /**
+     * Determine whether a task directory can be deleted.
+     *
+     * A task directory can be deleted if:
+     * - the task has completed
+     * - the task directory hasn't already been deleted
+     * - all of its publish outputs have been published
+     * - all of its process consumers are closed
+     * - all of its task consumers are completed
+     *
+     * @param task
+     */
+    private boolean canDeleteTask(TaskRun task) {
+        final taskState = tasks[task]
+        final processState = processes[task.processor.name]
+
+        taskState.completed
+            && !taskState.deleted
+            && taskState.publishOutputs.isEmpty()
+            && processState.consumers.every( p -> processes[p].closed )
+            && taskState.consumers.every( t -> tasks[t].completed )
+    }
+
+    /**
+     * Delete a task directory.
+     *
+     * @param task
+     */
+    private void deleteTask(TaskRun task) {
+        log.trace "[${task.name}] Deleting task directory: ${task.workDir.toUriString()}"
+
+        // delete task
+        threadPool.submit({
+            try {
+                FileHelper.deletePath(task.workDir)
```

Comment:
This looks like we remove the whole task working directory, which would also destroy the .command files. These files are used by the Seqera Platform and are generally useful for task provenance. Would it be sensible to consider whitelisting the Nextflow (and maybe Fusion) dotfiles from deletion?

Reply:
My plan was to save this provenance information in the `.nextflow` metadata cache, but right now that code is sitting in #3849. This information can also be preserved by using nf-prov

How else are they used by Seqera Platform?

Reply:
@robsyme I managed to move this PR to a plugin, so feel free to make an issue there if you want to continue this discussion about deleting the task directory.

I could add a config option to control whether to delete the task directory or only the outputs. Also, the upcoming workflow output DSL might make it very difficult to know when a task directory could be deleted, so it might become a moot point